core/src/apps/solana/sign_tx.py

```diff
@@ -167,4 +185,136 @@ def calculate_fee(transaction: Transaction) -> int:
+    - System program allocate with seed instruction (the space data parameter)
+    - Associated token account program create instruction (165 bytes for Token, ??? for Token22)
+    - Associated token account program create idempotent instruction (165 bytes for Token, ??? for Token22, might not allocate)
+    """
+    from .constants import (
+        SOLANA_ACCOUNT_METADATA_SIZE,
+        SOLANA_RENT_EXEMPTION_MULTIPLIER,
+        SOLANA_RENT_PER_BYTE_EPOCH,
+        SOLANA_TOKEN22_CONFIDENTIAL_TRANSFER_ACCOUNT_EXTENSION_SIZE,
+        SOLANA_TOKEN22_CONFIDENTIAL_TRANSFER_FEE_ACCOUNT_EXTENSION_SIZE,
+        SOLANA_TOKEN22_CPI_GUARD_EXTENSION_SIZE,
+        SOLANA_TOKEN22_IMMUTABLE_OWNER_EXTENSION_SIZE,
+        SOLANA_TOKEN22_MEMO_TRANSFER_EXTENSION_SIZE,
+        SOLANA_TOKEN22_NON_TRANSFERABLE_ACCOUNT_EXTENSION_SIZE,
+        SOLANA_TOKEN22_PAUSABLE_ACCOUNT_EXTENSION_SIZE,
+        SOLANA_TOKEN22_TRANSFER_FEE_AMOUNT_EXTENSION_SIZE,
+        SOLANA_TOKEN22_TRANSFER_HOOK_ACCOUNT_EXTENSION_SIZE,
+        SOLANA_TOKEN_ACCOUNT_SIZE,
+        SOLANA_TVL_HEADER_SIZE,
+    )
+    from .transaction.instructions import (
+        _ASSOCIATED_TOKEN_ACCOUNT_PROGRAM_ID,
+        _ASSOCIATED_TOKEN_ACCOUNT_PROGRAM_ID_INS_CREATE,
+        _ASSOCIATED_TOKEN_ACCOUNT_PROGRAM_ID_INS_CREATE_IDEMPOTENT,
+        _SYSTEM_PROGRAM_ID,
+        _SYSTEM_PROGRAM_ID_INS_ALLOCATE,
+        _SYSTEM_PROGRAM_ID_INS_ALLOCATE_WITH_SEED,
+        _SYSTEM_PROGRAM_ID_INS_CREATE_ACCOUNT,
+        _SYSTEM_PROGRAM_ID_INS_CREATE_ACCOUNT_WITH_SEED,
+        _TOKEN_2022_PROGRAM_ID,
+        _TOKEN_PROGRAM_ID,
+    )
+
+    def lookup_account(transaction: Transaction, account: Account):
+        if len(account) == 2:
+            return account[0]
+        else:
+            _, index, address_type = account
+            if address_type == AddressType.AddressRw:
+                return transaction.address_lookup_tables_rw_addresses[index]
+            else:
+                return transaction.address_lookup_tables_ro_addresses[index]
+
+    allocation_estimate = 0
+    for instruction in transaction.instructions:
+        if instruction.program_id == _SYSTEM_PROGRAM_ID and (
+            instruction.instruction_id == _SYSTEM_PROGRAM_ID_INS_CREATE_ACCOUNT
+            or instruction.instruction_id
+            == _SYSTEM_PROGRAM_ID_INS_CREATE_ACCOUNT_WITH_SEED
+            or instruction.instruction_id == _SYSTEM_PROGRAM_ID_INS_ALLOCATE
+            or instruction.instruction_id == _SYSTEM_PROGRAM_ID_INS_ALLOCATE_WITH_SEED
+        ):
+            allocation_estimate += (
+                instruction.parsed_data["space"] + SOLANA_ACCOUNT_METADATA_SIZE
+            )
+        elif instruction.program_id == _ASSOCIATED_TOKEN_ACCOUNT_PROGRAM_ID and (
+            instruction.instruction_id
+            == _ASSOCIATED_TOKEN_ACCOUNT_PROGRAM_ID_INS_CREATE
+            or instruction.instruction_id
+            == _ASSOCIATED_TOKEN_ACCOUNT_PROGRAM_ID_INS_CREATE_IDEMPOTENT
+        ):
```

Comment:
I guess `instruction.instruction_id in (..)` operator could be used here as well.